### ISSUE TITLE
Node GRPC memory leak

### ISSUE THREAD

**Issue by w4-sjcho** [grpc/grpc#4211]
I've created a sample client/server based on the node greeting example:
https://github.com/w4-sjcho/grpc-test

The client infinitely sends out request to server (one at a time), while the server returns a somewhat large response (100k). I've made the server to print out memory usage every 10 secs.

If you see the results:
```
$ node server.js
{"rss":32100352,"heapTotal":18061824,"heapUsed":9985712}
{"rss":82300928,"heapTotal":19093760,"heapUsed":7539352}
{"rss":114245632,"heapTotal":19093760,"heapUsed":9203120}
{"rss":170647552,"heapTotal":19093760,"heapUsed":10973264}
{"rss":206450688,"heapTotal":27482368,"heapUsed":8606624}
{"rss":210214912,"heapTotal":27482368,"heapUsed":10028640}
{"rss":243965952,"heapTotal":27482368,"heapUsed":11545720}
{"rss":301223936,"heapTotal":27482368,"heapUsed":13064544}
{"rss":358653952,"heapTotal":27482368,"heapUsed":14592936}
{"rss":405127168,"heapTotal":27482368,"heapUsed":8013472}
{"rss":408047616,"heapTotal":27482368,"heapUsed":9549824}
{"rss":414121984,"heapTotal":27482368,"heapUsed":11072176}
{"rss":418131968,"heapTotal":27482368,"heapUsed":12572664}
{"rss":438489088,"heapTotal":27482368,"heapUsed":14096648}
{"rss":495341568,"heapTotal":27482368,"heapUsed":15586968}
{"rss":515166208,"heapTotal":27482368,"heapUsed":8267008}
{"rss":515674112,"heapTotal":27482368,"heapUsed":9822464}
{"rss":515874816,"heapTotal":27482368,"heapUsed":11383728}
{"rss":520425472,"heapTotal":27482368,"heapUsed":12940736}
{"rss":575549440,"heapTotal":27482368,"heapUsed":14461304}
{"rss":598585344,"heapTotal":27482368,"heapUsed":7745752}
{"rss":608178176,"heapTotal":27482368,"heapUsed":9275504}
{"rss":608399360,"heapTotal":27482368,"heapUsed":10794704}
{"rss":609452032,"heapTotal":27482368,"heapUsed":12305632}
{"rss":657637376,"heapTotal":27482368,"heapUsed":13811912}
```
In 250 secs, the server memory (rss) increased from 32MB to 657MB, while heap usage isn't changed much.

Test env:
GRPC: Head version
OS: OSX 10.11
Node: 0.12.8

**Comment by w4-sjcho** [comment]
Please note that both server and client have leaks.

There are 2 kinds of objects that are leaked --- one is byte_buffer used during send message op.
I think it can be fixed with: https://github.com/w4-sjcho/grpc/commit/8b77d61d8ff204780770e8e2a13b586526f3ef3b

The other seems to be slice object (8.5kb according to Instruments) used when receiving data, but I can't find out where.

**Comment by w4-sjcho** [comment]
I think I found the other leak:
https://github.com/w4-sjcho/grpc/commit/5da98d461af9208f3b2630cd7339e017a2bd575d

With these 2 changes, both client and server works w/o leaks.

**Comment by murgatroid99** [comment]
This is very helpful, thank you for tracking this down. Can you send us a pull request with those changes?

**Comment by w4-sjcho** [comment]
I've sent out #4219

**Comment by w4-sjcho** [comment]
BTW, haven't checked but the php & ruby bindings might have the same leak described in:
https://github.com/w4-sjcho/grpc/commit/5da98d461af9208f3b2630cd7339e017a2bd575d

https://github.com/grpc/grpc/blob/release-0_11_1/src/ruby/ext/grpc/rb_byte_buffer.c#L65
https://github.com/grpc/grpc/blob/release-0_11_1/src/php/ext/grpc/byte_buffer.c#L74

**Comment by murgatroid99** [comment]
This was fixed by #4219.
